test/lit/passes/vacuum-func.wast

```diff
@@ -0,0 +1,60 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+;; RUN: wasm-opt %s --vacuum -all -S -o - | filecheck %s
+
+;; Tests vacuuming the entire body of a function. In that case we can ignore
+;; effects like a return or changes to locals.
+
+(module
+  ;; CHECK:      (func $optimizable (param $x i32)
+  ;; CHECK-NEXT:  (local $y i32)
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $optimizable (param $x i32)
+    (local $y i32)
+    ;; This entire function body can be optimized out. First, operations on
+    ;; locals are not observable once the function exits.
+    (local.set $x
+      (i32.const 1)
+    )
+    (local.set $y
+      (i32.const 2)
+    )
+    ;; Second, a return has no noticeable effect for the caller to notice.
+    (return)
+  )
+
+  ;; CHECK:      (func $partial (param $x i32)
+  ;; CHECK-NEXT:  (local $y i32)
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (local.get $x)
+  ;; CHECK-NEXT:   (unreachable)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $x
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $y
+  ;; CHECK-NEXT:   (i32.const 2)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (return)
+  ;; CHECK-NEXT: )
+  (func $partial (param $x i32)
+    (local $y i32)
+
+    ;; As above, but with this if added which adds extra possible effects. This
+    ;; prevents optimization.
+    ;; We could, however, optimize away the local operations TODO (we can also
```

Comment:
Don't we already have a pass that will remove unused `local.set`s? What more is there for vacuum to do here?

Reply:
We do, yeah, but there is actually a tricky case I'm worried about, which is why I wrote this (but I can delete it). Consider this code:
```js
function foo(i) {
  while (i < 1000) {
    i += 22;
    i *= 3;
  }
}
```
It seems obvious that code is not needed, since the result of the computation does not go anywhere. But none of our existing passes remove it, since the local `i` is both read and written. SSA form wouldn't help here AFAICT, we'd have a phi that has a use - itself - and so it seems used. What I suppose is necessary is a reachability analysis to see if the result of a computation "escapes" from the function (into either a return value or global state). Do you happen to know if that's what LLVM does, or if I'm missing something simpler?

Reply:
Makes sense, thanks. It would be good to update the comment to mention that it's more complicated cases where values locally appear used that needs the additional optimization.

I don't know specifically what LLVM does here, but my guess is that they have a more sophisticated data flow analysis that shows that the values computed here don't end up going anywhere. I hope that if we integrated the interpreter into GUFA we would be able to discover the same thing.